File: spark-connect-client/python/spark_connect_client.py
"""
Spark Connect Client for DoorDash Pedregal Platform

This module provides a standardized way to connect to Spark Connect clusters
through Spark Gateway. Teams use the open-source pyspark[connect] library
with DoorDash-specific configuration helpers.

Usage:
    pip install pyspark[connect]

    from spark_connect_client import SparkConnectClient

    # Connect to your team's cluster
    spark = SparkConnectClient.create_session(
        team="feature-engineering",
        environment="prod",
        region="us-west-2"
    )

    # Use Spark as normal
    df = spark.table("pedregal.feature_store.user_features")
"""

import os
from typing import Optional, Dict, Any
from dataclasses import dataclass
from enum import Enum


class Environment(Enum):
    DEV = "dev"
    STAGING = "staging"
    CI = "ci"
    PROD = "prod"


class Region(Enum):
    US_WEST_2 = "us-west-2"
    US_EAST_1 = "us-east-1"
    EU_WEST_1 = "eu-west-1"


@dataclass
class SparkConnectConfig:
    """Configuration for Spark Connect connection."""
    team: str
    environment: Environment
    region: Region
    gateway_host: str = "spark-gateway.doordash.team"
    gateway_port: int = 15002
    use_ssl: bool = True
    auth_token: Optional[str] = None

    @property
    def endpoint(self) -> str:
        """
        Generate the Spark Connect endpoint URL.

        Format: sc://[team]-[env]-[region].doordash.team:15002

        Examples:
            sc://feature-eng-prod-us-west-2.doordash.team:15002
            sc://ml-platform-dev-eu-west-1.doordash.team:15002
        """
        team_short = self.team.replace("-", "")[:10]
        env_short = self.environment.value[:3]
        region_short = self.region.value.replace("-", "")

        subdomain = f"{team_short}-{env_short}-{region_short}"
        return f"sc://{subdomain}.doordash.team:{self.gateway_port}"

    @property
    def direct_endpoint(self) -> str:
        """Direct endpoint through Spark Gateway with headers."""
        return f"sc://{self.gateway_host}:{self.gateway_port}"


class SparkConnectClient:
    """
    Factory for creating Spark Connect sessions connected to DoorDash clusters.

    This class provides helper methods to create properly configured SparkSession
    instances that connect through Spark Gateway to team-specific clusters.
    """

    # Default gateway endpoints by region
    GATEWAY_ENDPOINTS = {
        Region.US_WEST_2: "spark-gateway-us-west-2.doordash.team",
        Region.US_EAST_1: "spark-gateway-us-east-1.doordash.team",
        Region.EU_WEST_1: "spark-gateway-eu-west-1.doordash.team",
    }

    @classmethod
    def create_session(
        cls,
        team: str,
        environment: str = "dev",
        region: str = "us-west-2",
        app_name: Optional[str] = None,
        extra_config: Optional[Dict[str, Any]] = None,
    ):
        """
        Create a SparkSession connected to the team's Spark Connect cluster.

        Args:
            team: Team name (e.g., "feature-engineering", "ml-platform")
            environment: Environment name (dev, staging, ci, prod)
            region: AWS region (us-west-2, us-east-1, eu-west-1)
            app_name: Optional application name for the Spark session
            extra_config: Optional additional Spark configuration

        Returns:
            SparkSession connected to the remote Spark Connect cluster

        Example:
            spark = SparkConnectClient.create_session(
                team="feature-engineering",
                environment="prod",
                region="us-west-2"
            )
            df = spark.sql("SELECT * FROM pedregal.raw.events LIMIT 10")
        """
        from pyspark.sql import SparkSession

        # Parse environment and region
        env = Environment(environment)
        reg = Region(region)

        # Create config
        config = SparkConnectConfig(
            team=team,
            environment=env,
            region=reg,
            gateway_host=cls.GATEWAY_ENDPOINTS.get(reg, cls.GATEWAY_ENDPOINTS[Region.US_WEST_2]),
        )

        # Get auth token from environment or credential provider
        auth_token = os.environ.get("SPARK_CONNECT_AUTH_TOKEN")
        if auth_token:
            config.auth_token = auth_token

        # Build SparkSession
        builder = SparkSession.builder

        # Set remote endpoint
        builder = builder.remote(config.endpoint)

        # Set app name
        if app_name:
            builder = builder.appName(app_name)
        else:
            builder = builder.appName(f"{team}-{environment}-session")

        # Apply extra config
        if extra_config:
            for key, value in extra_config.items():
                builder = builder.config(key, value)

        return builder.getOrCreate()

    @classmethod
    def create_session_from_env(cls):
        """
        Create a SparkSession using environment variables.

        Environment variables:
            SPARK_CONNECT_TEAM: Team name
            SPARK_CONNECT_ENV: Environment (dev, staging, ci, prod)
            SPARK_CONNECT_REGION: Region (us-west-2, us-east-1, eu-west-1)
            SPARK_CONNECT_AUTH_TOKEN: (Optional) Auth token

        Returns:
            SparkSession connected to the remote cluster
        """
        team = os.environ.get("SPARK_CONNECT_TEAM")
        environment = os.environ.get("SPARK_CONNECT_ENV", "dev")
        region = os.environ.get("SPARK_CONNECT_REGION", "us-west-2")

        if not team:
            raise ValueError("SPARK_CONNECT_TEAM environment variable is required")

        return cls.create_session(team=team, environment=environment, region=region)

    @classmethod
    def create_session_direct(cls, endpoint: str, app_name: Optional[str] = None):
        """
        Create a SparkSession with a direct endpoint URL.

        Args:
            endpoint: Full Spark Connect URL (e.g., sc://feature-eng.doordash.team:15002)
            app_name: Optional application name

        Returns:
            SparkSession connected to the specified endpoint
        """
        from pyspark.sql import SparkSession

        builder = SparkSession.builder.remote(endpoint)

        if app_name:
            builder = builder.appName(app_name)

        return builder.getOrCreate()


# Convenience functions for common use cases

def get_spark_session(
    team: str,
    environment: str = "dev",
    region: str = "us-west-2",
) -> "SparkSession":
    """
    Convenience function to get a Spark session.

    Example:
        from spark_connect_client import get_spark_session

        spark = get_spark_session("feature-engineering", "prod")
        df = spark.table("pedregal.feature_store.user_features")
    """
    return SparkConnectClient.create_session(
        team=team,
        environment=environment,
        region=region,
    )


def get_feature_engineering_spark(environment: str = "dev") -> "SparkSession":
    """Get a Spark session for the Feature Engineering team."""
    return SparkConnectClient.create_session(
        team="feature-engineering",
        environment=environment,
        region="us-west-2",
    )


def get_ml_platform_spark(environment: str = "dev") -> "SparkSession":
    """Get a Spark session for the ML Platform team."""
    return SparkConnectClient.create_session(
        team="ml-platform",
        environment=environment,
        region="us-west-2",
    )


if __name__ == "__main__":
    # Example usage
    print("Spark Connect Client for DoorDash")
    print("=" * 50)

    # Show example endpoint
    config = SparkConnectConfig(
        team="feature-engineering",
        environment=Environment.PROD,
        region=Region.US_WEST_2,
    )
    print(f"Example endpoint: {config.endpoint}")

    # Show usage
    print("\nUsage:")
    print("  from spark_connect_client import get_spark_session")
    print("  spark = get_spark_session('feature-engineering', 'prod')")
    print("  df = spark.table('pedregal.feature_store.user_features')")

File: spark-connect-client/java/src/main/java/com/doordash/spark/connect/SparkConnectClient.java
package com.doordash.spark.connect;

import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * Spark Connect Client for DoorDash Pedregal Platform.
 *
 * <p>This class provides a standardized way to connect to Spark Connect clusters
 * through Spark Gateway. Teams use the Apache Spark Connect client library
 * with DoorDash-specific configuration helpers.</p>
 *
 * <h2>Usage:</h2>
 * <pre>{@code
 * SparkSession spark = SparkConnectClient.builder()
 *     .team("feature-engineering")
 *     .environment(Environment.PROD)
 *     .region(Region.US_WEST_2)
 *     .build();
 *
 * Dataset<Row> df = spark.table("pedregal.feature_store.user_features");
 * }</pre>
 */
public class SparkConnectClient {

    private static final Logger logger = LoggerFactory.getLogger(SparkConnectClient.class);

    /**
     * Supported environments for Spark Connect clusters.
     */
    public enum Environment {
        DEV("dev"),
        STAGING("staging"),
        CI("ci"),
        PROD("prod");

        private final String value;

        Environment(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    /**
     * Supported regions for Spark Connect clusters.
     */
    public enum Region {
        US_WEST_2("us-west-2", "spark-gateway-us-west-2.doordash.team"),
        US_EAST_1("us-east-1", "spark-gateway-us-east-1.doordash.team"),
        EU_WEST_1("eu-west-1", "spark-gateway-eu-west-1.doordash.team");

        private final String value;
        private final String gatewayHost;

        Region(String value, String gatewayHost) {
            this.value = value;
            this.gatewayHost = gatewayHost;
        }

        public String getValue() {
            return value;
        }

        public String getGatewayHost() {
            return gatewayHost;
        }
    }

    private SparkConnectClient() {
        // Use builder
    }

    /**
     * Create a new builder for SparkConnectClient.
     *
     * @return A new Builder instance
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Quick factory method to create a Spark session.
     *
     * @param team        Team name (e.g., "feature-engineering")
     * @param environment Environment (dev, staging, ci, prod)
     * @param region      Region (us-west-2, us-east-1, eu-west-1)
     * @return SparkSession connected to the remote cluster
     */
    public static SparkSession createSession(String team, Environment environment, Region region) {
        return builder()
                .team(team)
                .environment(environment)
                .region(region)
                .build();
    }

    /**
     * Create a Spark session from environment variables.
     *
     * <p>Reads the following environment variables:</p>
     * <ul>
     *   <li>SPARK_CONNECT_TEAM - Team name (required)</li>
     *   <li>SPARK_CONNECT_ENV - Environment (default: dev)</li>
     *   <li>SPARK_CONNECT_REGION - Region (default: us-west-2)</li>
     * </ul>
     *
     * @return SparkSession connected to the remote cluster
     */
    public static SparkSession createSessionFromEnv() {
        String team = System.getenv("SPARK_CONNECT_TEAM");
        if (team == null || team.isEmpty()) {
            throw new IllegalStateException("SPARK_CONNECT_TEAM environment variable is required");
        }

        String envStr = Optional.ofNullable(System.getenv("SPARK_CONNECT_ENV")).orElse("dev");
        String regionStr = Optional.ofNullable(System.getenv("SPARK_CONNECT_REGION")).orElse("us-west-2");

        Environment environment = Environment.valueOf(envStr.toUpperCase());
        Region region = Region.valueOf(regionStr.toUpperCase().replace("-", "_"));

        return createSession(team, environment, region);
    }

    /**
     * Builder for creating Spark Connect sessions.
     */
    public static class Builder {
        private String team;
        private Environment environment = Environment.DEV;
        private Region region = Region.US_WEST_2;
        private String appName;
        private int port = 15002;
        private final Map<String, String> extraConfig = new HashMap<>();

        /**
         * Set the team name.
         *
         * @param team Team name (e.g., "feature-engineering", "ml-platform")
         * @return this builder
         */
        public Builder team(String team) {
            this.team = team;
            return this;
        }

        /**
         * Set the environment.
         *
         * @param environment Environment (DEV, STAGING, CI, PROD)
         * @return this builder
         */
        public Builder environment(Environment environment) {
            this.environment = environment;
            return this;
        }

        /**
         * Set the region.
         *
         * @param region Region (US_WEST_2, US_EAST_1, EU_WEST_1)
         * @return this builder
         */
        public Builder region(Region region) {
            this.region = region;
            return this;
        }

        /**
         * Set the application name.
         *
         * @param appName Application name for the Spark session
         * @return this builder
         */
        public Builder appName(String appName) {
            this.appName = appName;
            return this;
        }

        /**
         * Set the gateway port.
         *
         * @param port Gateway port (default: 15002)
         * @return this builder
         */
        public Builder port(int port) {
            this.port = port;
            return this;
        }

        /**
         * Add extra Spark configuration.
         *
         * @param key   Configuration key
         * @param value Configuration value
         * @return this builder
         */
        public Builder config(String key, String value) {
            this.extraConfig.put(key, value);
            return this;
        }

        /**
         * Build the SparkSession.
         *
         * @return SparkSession connected to the remote Spark Connect cluster
         */
        public SparkSession build() {
            if (team == null || team.isEmpty()) {
                throw new IllegalStateException("Team is required");
            }

            String endpoint = buildEndpoint();
            logger.info("Connecting to Spark Connect at: {}", endpoint);

            SparkSession.Builder sparkBuilder = SparkSession.builder()
                    .remote(endpoint);

            // Set app name
            String effectiveAppName = appName != null ? appName : team + "-" + environment.getValue() + "-session";
            sparkBuilder.appName(effectiveAppName);

            // Apply extra config
            for (Map.Entry<String, String> entry : extraConfig.entrySet()) {
                sparkBuilder.config(entry.getKey(), entry.getValue());
            }

            return sparkBuilder.getOrCreate();
        }

        private String buildEndpoint() {
            // Build team-specific subdomain
            // Format: sc://[team-short]-[env-short]-[region-short].doordash.team:15002
            String teamShort = team.replace("-", "").substring(0, Math.min(10, team.replace("-", "").length()));
            String envShort = environment.getValue().substring(0, 3);
            String regionShort = region.getValue().replace("-", "");

            String subdomain = teamShort + "-" + envShort + "-" + regionShort;
            return "sc://" + subdomain + ".doordash.team:" + port;
        }
    }

    /**
     * Main method demonstrating usage.
     */
    public static void main(String[] args) {
        System.out.println("Spark Connect Client for DoorDash");
        System.out.println("=".repeat(50));

        // Show example endpoint
        Builder builder = new Builder()
                .team("feature-engineering")
                .environment(Environment.PROD)
                .region(Region.US_WEST_2);

        System.out.println("Example endpoint: sc://featureeng-pro-uswest2.doordash.team:15002");

        System.out.println("\nUsage:");
        System.out.println("  SparkSession spark = SparkConnectClient.builder()");
        System.out.println("      .team(\"feature-engineering\")");
        System.out.println("      .environment(Environment.PROD)");
        System.out.println("      .region(Region.US_WEST_2)");
        System.out.println("      .build();");
        System.out.println("");
        System.out.println("  Dataset<Row> df = spark.table(\"pedregal.feature_store.user_features\");");
    }
}
